[IMP] web_editor, website, *: allow columns on mobile snippets

robinlej opened the pull request:
*: mass_mailing

This PR introduces options for the flex column layout and the
horizontal resizing to be used on mobile and be independent from
the desktop layout (commit 3).

In the other commits, it also:
- introduces 2 utility functions `isMobile()` and `isMobileView()`,
respectively in website and web_editor, to check whether the view is
either mobile or desktop. (commit 1)
- moves logic from the method `_computeWidgetVisibility()` in the main
`SnippetOptionsWidget` to the overriding method in the `SnippetMove`
sub-widget instead. (commit 2)
- adds a test for the behavior of the column layout. (commit 3)

In particular, commit 3 makes it possible to:
- Have different number of columns on mobile and on desktop.
- Use different column widths and offsets for mobile and desktop.
- Reorder columns independently.

The behavior expected from the columns is:
- In the editor panel, Cols indicate the number of columns per row on
both mobile and desktop. The option is conditional of your environment:
you edit the number for the screen size of your current resolution.
- When columns have different sizes (either because it is the default
behavior of the snippet, or because the user resized some), the counter
should read "Custom". Same when the user changed some offsets.
- If manual modifications were made on width and offset, they are reset
on the current display if the number of elements is updated. This was
decided because an old, specific layout with custom offsets / width
doesn't work well with a different number of elements. (This is also the
default behavior before this PR with desktop-only modifications.)

The expected behavior when reordering is:
- On mobile, it should only affect the mobile layout.
- On desktop, it should affect both layouts.

In order to have a coherent feature on both mobile and desktop as well
as correct some nonideal but non blocking behaviors, the commit also
changes the following behaviors:
- When setting a columns count lower than the current number of items in
the `.row` container, extra items are now wrapped on the next flex-rows
(still within the container) instead of being deleted.
- When going from e.g. 3 to 5 columns (and as many items), the change
happens all at once instead of each item being visibly added one by one
on the next row, then brought back to the first row.
- The columns count automatically updates when resizing an item, instead
of updating only after you click somewhere else on the snippet.

task-3097045

robodoo commented:
[Pull request status dashboard](https://mergebot.odoo.com/odoo/odoo/pull/117562).

Guillaume-gdi commented on addons/web_editor/static/src/js/editor/snippets.options.js:
```diff
@@ -4791,19 +4799,26 @@ registry['sizing_x'] = registry.sizing.extend({
                 }
-                this.$target.attr('class', this.$target.attr('class').replace(/\s*(offset-xl-|offset-lg-|col-lg-)([0-9-]+)/g, ''));
+                const offsetColRegex = new RegExp(`\\b(offset-${resolutionModifierWithXL}|col-${resolutionModifier})(\\d{1,2})`, 'g');
+                this.$target.attr('class', this.$target.attr('class').replace(offsetColRegex, ''));
```
We try to get rid of JQuery, you can take the opportunity to replace it by a pure JS code :+1:

Guillaume-gdi commented on addons/web_editor/static/src/js/editor/snippets.options.js:
```diff
@@ -5251,50 +5282,105 @@ registry.layout_column = SnippetOptionWidget.extend({
      */
-    _updateColumnCount: async function ($row, count) {
-        if (!count) {
+    _updateColumnCount: async function ($row, nbColumns) {
```
As you rewrite this, you can do it in the new way
`async _updateColumnCount(...) {`

Guillaume-gdi commented on addons/web_editor/static/src/js/editor/snippets.options.js:
```diff
@@ -5251,50 +5282,105 @@ registry.layout_column = SnippetOptionWidget.extend({
+        const itemOffset = this._getItemOffset(itemsOnLastRow, itemSize, nbColumns);
+        const itemClass = `col${resolutionModifier}-${itemSize}`;
+        const mobileItemSize = !isMobile ? this._getItemSize($items[0], true) : itemSize;
+        _.each($items, function (item) {
```
We try to get rid of undescore, you can itterate over your items with pure JS

Guillaume-gdi commented on addons/web_editor/static/src/js/editor/snippets.options.js:
```diff
@@ -5251,50 +5282,105 @@ registry.layout_column = SnippetOptionWidget.extend({
+        _.each($items, function (item) {
+            if (sameNbColumnsMobileDesktop) {
+                // (?![\w-]): means \b (word boundary) except -. This is necessary to avoid matching
+                //              partial classes (typically: 'col' inside 'col-lg-2' while looking for
```
I'm not a huge fan of the 14 blank space after `//`, one is enough :)

Guillaume-gdi commented on addons/web_editor/static/src/js/editor/snippets.options.js:
```diff
@@ -5331,6 +5417,80 @@ registry.layout_column = SnippetOptionWidget.extend({
+        const items = this.$target[0].querySelector(':scope > .row')?.children;
+        if (isMobile && items) {
+            const itemSize = this._getItemSize(items[0], isMobile);
+            if (itemSize * items.length > 12 ) {
```
You can remove the blank space after 12

Guillaume-gdi commented on addons/web_editor/static/src/js/editor/snippets.options.js:
```diff
@@ -5331,6 +5417,80 @@ registry.layout_column = SnippetOptionWidget.extend({
+     * @param {integer} itemOffset - offset to add to the item
+     */
+    _addMobileOffset($items, itemsOnLastRow, itemOffset) {
+        const targetItem = $items[$items.length - itemsOnLastRow];
```
As `$items[$items.length - itemsOnLastRow]` is a HTML node, the variable name should finish by `El`

Guillaume-gdi reviewed (changes requested):
Hello, here is a first code review, I didn't check the regex for the moment.

In addition to the comments below, some lines of code are too long, the lines of comments can not exceed the 80th col and the lines of code can not exceed the 100th col.

I did a small test of the feature and I noticed this behavior:
- Drop a Image - Text block
- Toggle the mobile view
- Set the nb of col to 2
- Reduce the size of the right col

=> It's also reduced in desktop :grimacing: 

If I understood correctly, this part of the spec says that col size should be computed for mobile only in this case or am I wrong ?
![image](https://user-images.githubusercontent.com/78849981/230085639-50c1728c-78ff-4d7f-a904-975249664266.png)

Courage :muscle:

Guillaume-gdi commented on addons/web_editor/static/src/js/editor/snippets.options.js:
```diff
@@ -5250,51 +5279,88 @@ registry.layout_column = SnippetOptionWidget.extend({
+        [...columnEls].forEach(columnEl => {
+            columnEl.className = columnEl.className
+                .replace(resolutionRegex, '')
+                .replace(/\boffset-lg-0\b/, '');
```
Is that just to remove the `offset-lg-0` class ? If yes you can just do `columnEl.classList.remove('offset-lg-0')`

robinlej replied:
That's when you reorganize the code several times and in the end you don't see the obvious. 🤦

robinlej commented on addons/web_editor/static/src/js/common/utils.js:
```diff
@@ -408,6 +408,23 @@ function _getColorClass(el, colorNames, prefix) {
+ * @param {Object} self - context object ("this")
+ * @returns {boolean}
+ */
+function _isMobile(self) {
+    let isMobile;
+    self.trigger_up('service_context_get', {
+        callback: (ctx) => {
+            isMobile = ctx['isMobile'];
+        },
+    });
+
+    return isMobile;
+}
```
What do you mean? I replaced every occurrence where I used it.

Guillaume-gdi replied:
The purpose of a utility function is (among other things) to prevent having duplicated code in several places. Here you have added this new utility, it's great but you should make sure that we use this utility in the whole codebase. So you also have to adapt the `toggleDeviceVisibility()` of `DeviceVisibility` option, the `_updateInvisibleDOM` of the `SnippetsMenu` widget

robinlej commented on addons/web_editor/static/src/js/editor/snippets.options.js:
```diff
@@ -5331,6 +5397,111 @@ registry.layout_column = SnippetOptionWidget.extend({
+                        return el;
+                    }
+                })
+                .filter(el => el);
```
.map returns undefined for unused elements, so `.filter(el => el)` filters them out. :)

Guillaume-gdi replied:
I think it's a bad idea to use map to get `firstRowEls` so. You itterate over elements unnecessarily. You could compute it with a while or a for loop. I think the code will be shorter and easier to read :+1:

Guillaume-gdi replied:
Re thinking about it, I would clearly go with a `filter`. Map() is to map each value to another one. Here it's not what you have to do, you need to **filter** the `columnEls` to get the `firstRowEls`.

robinlej replied:
Very true.

robinlej commented on addons/web_editor/static/src/js/editor/snippets.options.js:
```diff
@@ -4763,47 +4762,76 @@ registry['sizing_x'] = registry.sizing.extend({
+
+            const beginOffset = Number(beginClass.match(offsetRegexWithXL)[1] || 0);
+
+            const targetEl = this.$target[0];
```
It's actually used 11 times in the if / else that follows, I think it's cleaner than repeating `this.$target[0]`. But I can remove it if you really think otherwise.

Guillaume-gdi replied:
Ok, I suppose it's a matter of taste

robinlej commented on addons/web_editor/static/src/js/editor/snippets.options.js:
```diff
@@ -5250,51 +5279,88 @@ registry.layout_column = SnippetOptionWidget.extend({
-        } else {
-            var self = this;
-            for (const el of $row.children().slice(count)) {
+        } else if (itemsDelta < 0 && prevNbRows < 2) {
```
I'll add a comment in the code, but it's part of the specs: "when you decrease the number, we delete items only if everything was in one row."
Basically, if we had more than 1 row, we keep all the items and just reorganize the columns on the right amount of rows.

Guillaume-gdi commented on addons/website/static/src/js/editor/snippets.options.js:
The commit title doesn't respect the convention, you have to mention all the changes addons. Don't forget that an important part of the development is self review :wink:

Guillaume-gdi commented on addons/web_editor/static/src/js/editor/snippets.options.js:
```diff
@@ -4762,47 +4761,76 @@ registry['sizing_x'] = registry.sizing.extend({
         this.grid = {
-            e: [_.map(gridE, v => ('col-lg-' + v)), _.map(gridE, v => width / 12 * v), 'width'],
-            w: [_.map(gridW, v => ('offset-lg-' + v)), _.map(gridW, v => width / 12 * v), 'margin-left'],
+            e: [
+                _.map(gridE, v => (`col-${resolutionModifier}${v}`)),
+                _.map(gridE, v => width / 12 * v),
+                'width',
+            ],
+            w: [
+                _.map(gridW, v => (`offset-${resolutionModifier}${v}`)),
+                _.map(gridW, v => width / 12 * v),
+                'margin-left',
+            ],
```
You can get rid of Underscore here

Guillaume-gdi commented on addons/web_editor/static/src/js/editor/snippets.options.js:
```diff
@@ -5249,51 +5278,94 @@ registry.layout_column = SnippetOptionWidget.extend({
      * @private
-     * @param {jQuery} $row - the row in which to update the columns
-     * @param {integer} count - positif to add, negative to remove
+     * @param {Element} rowEl - the row in which to update the columns
```
`HTMLElement` would be better here :+1:

Guillaume-gdi reviewed (changes requested):
I did just a quick check, do not hesitate to ask another review to the new helpers of the week if you need so before going to the next steps :slightly_smiling_face:

qsm-odoo commented on addons/web_editor/static/src/js/editor/snippets.options.js:
```diff
@@ -4763,47 +4762,76 @@ registry['sizing_x'] = registry.sizing.extend({
                     colSize = 1;
                     offset = beginOffset + beginCol - 1;
                 }
-                this.$target.attr('class', this.$target.attr('class').replace(/\s*(offset-xl-|offset-lg-|col-lg-)([0-9-]+)/g, ''));
```
The regex seemed already wrong actually, not worth a fix in stable I guess but time to fix it :)

qsm-odoo commented on addons/web_editor/static/src/js/editor/snippets.options.js:
```diff
@@ -5250,51 +5279,99 @@ registry.layout_column = SnippetOptionWidget.extend({
      */
-    _updateColumnCount: async function ($row, count) {
-        if (!count) {
+    async _updateColumnCount(rowEl, nbColumns) {
+        // itemsDelta: positive to add, negative to remove
+        const itemsDelta = nbColumns - rowEl.children.length;
+        const isMobile = weUtils.isMobile(this);
+        const mobileColumns = this._getNbColumns(true);
```
Before digging deeper in this implementation, I'd like to ask @pierre-pde if it is really what he wants functionally,  there are many thing which are weird to me:
- Set up 3 column in desktop, it shows "3", go to mobile, make two of them side by side: it show 2 if you did it with the first 2 but 1 if you did it with the last 2.
- Why is the logic different between desktop and mobile:
    - Desktop: choose 3 columns, resize one so it wraps over 2 lines -> it still shows 3
    - Mobile: choose 3 columns, resize one so it wraps over 2 lines -> it shows 2 (or 1 as explained above)
- Bug: desktop choose 3 columns, resize the right one to make it bigger: does not work immediately (no issue in standard master)
- IMO, the order you set up on desktop/mobile should stay (kinda like what we do with grid). If I set up 3 columns in desktop: image - text -image, then in mobile set up the 2 images to be side-by-side and the text below... it should not impact the desktop order IMO. The logic may not be very easy but there is a way to do it thanks to order-XXX classes.

I think there is more to test and even more to do here :wink:

qsm-odoo commented on addons/web_editor/static/src/js/editor/snippets.options.js:
```diff
@@ -5250,51 +5279,99 @@ registry.layout_column = SnippetOptionWidget.extend({
+            }
+        }
+
+        [...columnEls].forEach(columnEl => {
```
Especially when you have to do `[...myStuffWhichIsNotReallyAnArray]`, I think `for (const columnEl of columnEls) {` is a prettier syntax (first the temporary variable can be a `const` but it also can be "read" naturally "Pour chaque column parmi ces columns, do...". Feel free to disagree, if you prefer that way, it's ok :wink:

qsm-odoo commented on addons/website/static/src/snippets/s_color_blocks_2/000.scss:
```diff
@@ -23,11 +23,6 @@
         padding-top: 8vw; // A flex item cannot have % padding top and bottom (even if it works on chrome)
         padding-bottom: 8vw; // Solution is vw units but we keep 8% as a fallback
     }
-    @include media-breakpoint-down(lg) {
-        [class*="col-lg-"] {
-            flex: 1 1 100%;
-        }
-    }
```
Sure you are not breaking old snippets removing this rule? (real question)

qsm-odoo reviewed (changes requested):
I'll dig deeper after @pierre-pde feedbacks and following adaptations :wink: But intuitively, I am worried +248-64 is too big of a diff for the current feature (I may be wrong) so feel free to ignore that comment for now if you are sure everything should be as it is :+1:

robinlej commented on addons/web_editor/static/src/js/editor/snippets.options.js:
```diff
@@ -4750,8 +4749,8 @@ registry['sizing_x'] = registry.sizing.extend({
-            var _class = this.$target.attr('class').replace(/\s*(offset-xl-|offset-lg-)([0-9-]+)/g, '');
-            this.$target.attr('class', _class);
+            this.$target[0].className =
+                this.$target[0].className.replace(/\s*offset-(xl-|lg-)?([0-9-]+)/g, '');
```
Since it's the same problem with `\b`, I'll make it bullet proof with `(?<!\S)` and close it with `(?!\S)`, and modify my other regex too.

qsm-odoo replied:
Why not `(\s+|^)<your_class_regex>(\s+|$)` maybe ? (not even sure what your regexes stand for ^^). Otherwise maybe avoiding the "complex" part of the regex is a good solution too, something like:
```javascript
classList.remove(...classList.filter(cls => cls.match(/offset-(xl-|lg-)?[0-9]+/)));
```
(something like that)
Honestly it maybe be more readable. But a solo regex is good too. But if you use `(?<!\S)`, make sure to comment what it is because it is not straightforward (to me at least).

qsm-odoo commented on addons/web_editor/static/src/js/editor/snippets.editor.js:
```diff
@@ -3651,6 +3646,14 @@ var SnippetsMenu = Widget.extend({
+    /**
+     * Hook for website.
+     *
+     * @returns {bool} - false
```
```suggestion
     * @returns {boolean}
```

qsm-odoo commented on addons/web_editor/static/src/js/editor/snippets.editor.js:
```diff
@@ -3651,6 +3646,14 @@ var SnippetsMenu = Widget.extend({
+     * @returns {bool} - false
+     */
+    _isMobile() {
+        return false;
```
Would not it make sense to call the web_editor util you made actually? :shrug: It will return "false" in all supported usecase anyway but even when it returns "true", it won't hurt. It will actually be consistent... this code about invisible snippets does not make sense for mass mailing anyway (for now?).

qsm-odoo commented on addons/website/static/src/js/utils.js:
```diff
@@ -416,6 +416,24 @@ function generateGMapLink(dataset) {
+/**
+ * Checks if the current device is mobile.
+ *
+ * @private
```
Remove private here too

qsm-odoo commented on addons/website/static/src/js/utils.js:
```diff
@@ -416,6 +416,24 @@ function generateGMapLink(dataset) {
 }
 
+/**
+ * Checks if the current device is mobile.
```
Comment could be better: "Checks if the edited content is currently previewed as in a mobile device"?

qsm-odoo commented on addons/website/static/src/js/utils.js:
```diff
@@ -416,6 +416,24 @@ function generateGMapLink(dataset) {
+        },
+    });
+
+    return isMobile;
```
Actually maybe `isMobile ? true : <call web_editor utils>` ? We don't normally fully support editing the website from a mobile device... but I guess we will be going in the direction that we could at some point... would justify a bit more having an util for this. Either do that or remove this util entirely (and do what it does in the _isMobile override directly).

qsm-odoo commented on addons/web_editor/static/src/js/editor/snippets.options.js:
```diff
@@ -1043,7 +1044,7 @@ const SelectUserValueWidget = BaseSelectionUserValueWidget.extend({
             }
         } else {
-            textContent = this.PLACEHOLDER_TEXT;
+            textContent = _t("Custom");
```
Remind/Explain to me why this change was not about changing the value of `this.PLACEHOLDER_TEXT`, I don't remember/understand.

robinlej replied:
The `PLACEHOLDER_TEXT` constant was added in [b7b05a2d1c1d5](https://github.com/odoo/odoo/commit/b7b05a2d1c1d5), around the time I reworked on it after your 1st review, and consistently changed the text from "/" to "None" wherever it was used. (* except on the line `if (this.menuTogglerEl.textContent === '/') {` where it's apparently been forgotten.)
To my understanding, the "Custom" text we agreed on shouldn't be used wherever the PLACEHOLDER_TEXT is set (e.g. togglers where no value is selected should remain "None" - see for instance when you drop a "Dynamic snippet": filter & template are "None" by default, this depends on the placeholder).

qsm-odoo replied:
@robinlej 

> * except on the line if (this.menuTogglerEl.textContent === '/') { where it's apparently been forgotten

Did you make a PR to fix that? :+1: 

> togglers where no value is selected should remain "None"

Well that's the point :wink: Your PR is changing exactly that by "Custom". I tested the dynamic snippet to see what you were talking about (I used the "Products" snippet), it's exactly broken like I thought it would:
![image](https://github.com/odoo/odoo/assets/10338094/7415140f-1f51-4803-a84f-ab91f4b2713c)
No tags is selected yet and it shows "Custom", it does not make sense :wink: I don't think we ever agreed that all select that use a non-existing value in the selection list shows "Custom": this change should probably not be about a generic change on the widget but a parameter used by your specific select only, no?

qsm-odoo commented on addons/web_editor/static/src/js/editor/snippets.options.js:
```diff
@@ -4784,47 +4795,77 @@ registry['sizing_x'] = registry.sizing.extend({
      */
     _onResize: function (compass, beginClass, current) {
+        const targetEl = this.$target[0];
+        const isMobileView = weUtils.isMobileView(this.$target[0]);
```
Use targetEl which is just above? :)

qsm-odoo commented on addons/web_editor/static/src/js/common/column_layout_mixin.js:
```diff
@@ -0,0 +1,144 @@
+        const colSize = parseInt(columnEls[0].className.match(colRegex)?.[1] || 12);
+
+        const isCustomized =
+            columnEls[0].parentElement.classList.contains(`o_we_${screenType}-size-custom`);
```
- Don't mix _ and - in class names
- Don't build class name when you are not forced to: `isMobile ? o_we_mobile_custom_size : ...`. Eases grep and reduces mistakes.
- This class should be documented somewhere. Either here or probably rather near the code that adds it. I actually remember we spoke about it but I can't remember what was decided or why it is needed: a comment in the code should :+1:

robinlej replied:
These classes are needed because controlling whether all items have the same size is not enough with "weirder" cases, such as when the offsets have been meddled with (e.g. all items are `col-4`, but the 1st item has an `offset-2` set by the user: it should show "custom", not "3" columns). Likewise, I can't simply control the offsets to know if they've been altered, because there are default offsets, for instance with 5 columns (`offset-1` on the 1st, `offset-5` on the 6th), or with 3 columns and 5 items, and no easy way to differentiate them from custom ones.

But actually, this also means that the `allColSizesEqual` variable just after this seems useless after all, because if they're not equal, they must be customized.

qsm-odoo commented on addons/web_editor/static/src/js/editor/snippets.options.js:
```diff
@@ -3940,6 +3942,14 @@ const SnippetOptionWidget = Widget.extend({
+            // On mobile, items' reordering is independent from desktop inside
+            // a snippet (left or right), not at a higher level (up or down).
+            if (moveLeftOrRight && isMobileView &&
+                    ColumnLayoutMixin._getItemMobileOrder(this.$target[0])) {
```
I know it was like that before but what the hell is that `_computeWidgetVisibility` override doing in the main SnippetOptionWidget definition? :open_mouth:  Should not it be in the `SnippetMove` specific one? :thinking: Which would allow here to not call a private mixin method like that but just do `this._getItemMobileOrder`?

If yes, make a dedicated commit about moving this `_computeWidgetVisibility` definition out of here

qsm-odoo commented on addons/web_editor/static/src/js/editor/snippets.options.js:
```diff
@@ -5271,51 +5314,129 @@ registry.layout_column = SnippetOptionWidget.extend({
+        const isMobile = weUtils.isMobileView(this.$target[0]);
+        const prevNbColumns = this._getNbColumns(rowEl.children, isMobile);
+        const prevNbRows = !isMobile ? 1 : Math.ceil(rowEl.children.length / prevNbColumns);
+        // itemsDelta: positive to add, negative to remove.
+        // - Remove only if it was already on one row (otherwise, keep all
+        // items and reorganize the columns on the right amount of rows).
+        // - Add if there are more columns than elements.
+        const itemsDelta = prevNbRows === 1 || nbColumns > rowEl.children.length ?
+            nbColumns - rowEl.children.length : 0;
+
+        if (!itemsDelta && nbColumns === prevNbColumns) {
```
It's not clear enough how `prevNbColumns` can be different from `rowEl.children.length` here and how `!itemsDelta` can be different from `nbColumns === prevNbColumns` (I understand because I know the task but just reading the comment, I am actually confused).

qsm-odoo reviewed (changes requested):
Complex PR to review, if you made a mistake somewhere I am pretty sure I won't see it: I did not expect the latest requested changes to complexify the code that much (but this may be normal :grimacing:).

Here are (some superficial for the most part) requested changes. After that I'll try having a deeper look which will probably more be about deep testing this myself again.

Thanks :pray:

robinlej commented:
@qsm-odoo the changes have been made.

Btw, if I remember well, there were 2 main factors to explain the complexity for this second round:
1) the order that should be handled apart on mobile and on desktop: it implies quite a lot of checks here and there, and particularly it implies the use of the mixin to manipulate everything from `SnippetMove`, as well as a whole section on "what to do when you switch the order of two elements with different offsets, while the offsets might differ on mobile/desktop".
2) the fact that "Custom" should be used on desktop added logic to check those cases and logic to check when it has or doesn't have an impact on the rest.

qsm-odoo reviewed (changes requested):
Some superficial requested changes but important to do first if I want to test this properly. Also, when I ask for changes, when possible (do lose too much time if it is not easy), try not to push-force the branch multiple times and rebasing at the same time. If you do, I can't see what changes you made and I am forced to take a look at everything again. Ideally, after requested changes:
1. Work on the request changes
2. When it's time to push (end of the day / changes finished):
    a. First rebase the remote branch and push-force it
    b. Add your changes to the rebased branch and push-force again
=> After this Github now shows the diff of the changes that you made
3. You forgot something and have to push-force again: of course push-force again. In that case, the ideal would be to tell me the commit hash of the latest rebased before request changes, that way just a `git diff last_rebase..new_head` would give me your whole diff :+1: 

(Again, not to worry if the work was big and many changes were made, most of the time, I re-review the whole thing anyway but in this case, I asked myself what you changed in some parts and it's not easy to see).

I did a quick testing and found 2 bugs and weird things (that I think are worth re-speaking about with @pierre-pde), especially since (sorry): I am worried about the +500 LOC for a confusing (IMO) and not flawless feature.

First bug:
- Enter mass mailing edition
- Add a "Columns" snippet
- Resize a column
=> Instead of "Custom", I have "None". The option code was not adapted in mass_mailing.
(Note: small bug even in website, this goes to "Custom"/"None" only after a new column is clicked, which is strange... should not it update immediately (maybe it is a generic bug about column resizing not update the UI? which would actually reveal more problems I think if that's that) ?)

Second bug (tested on Firefox, confirmed it does occur only in this branch):
- Enter mass mailing edition
- Add a "Columns" snippet
- Click on the *middle* column
- Choose "4" columns instead of "3" in the sidebar
- Reduce the size of the current column (the second one) of one step
- Hit CTRL + SHIFT + Print screen (yes those are weird steps)
=> Traceback

Weird things that may compromise this feature (still the amount of code needed here is already a big worry to me... in general when a "small" (= for the user) feature requires hundres of LOC, I'd say something is wrong (either the code is not good enough or the spec should be changed... probably the spec in this case?). @pierre-pde When you compare what we have now and what we have after this task, do you think the users are less confused / more happy (worth of 500LOC more? ~5% more given all options combined)... I am not so sure anymore (but I can be wrong).

1. Enter edit mode, go on mobile edition
2. Add a columns snippet => 3 "columns", all below each other, the sidebar displays "1" (fine, there 3 things but displayed as 1 column)
3. Switch to "2" columns => There are now 3 things displayed on two columns and the second row is one item which is centered... first: is the code to center that column worth it? second: should not the more natural thing be about creating an extra fourth item to fill the row or removing the 3rd items to really have only 2 columns? Anything is fine here I guess... but nothing seems consistent with the rest to me, see next steps.
4. Switch back to "1" column => You are back as at step (2), nice !
5. Switch now to "4" columns => There are now an extra thing displayed as one row of 4 columns, nice... but in the end, > 3 columns on mobile is always ugly, maybe we should not even allow it?
6. Switch back to "1" column, "4" was really ugly, I want what I had back... well choosing "1" this time does not reset back to the way it was like it did at step (4), I just have one item left.

Combine this flow with desktop switching, re-ordering the items, and the display of "Custom" columns... I think we lost ourselves in these specs. Sorry @robinlej I am sure your code is ok and I don't want to throw it away but I think there is a problem here (I may be wrong). In any case, don't hesitate to voice concerns sooner (hitting a few hundreds lines for a small feature is a huge warning on its own but combined with the fact we don't have a perfect feature IMO... it should have raised more concerns sooner).

Another big worry: combine that with grid mode. Before this branch: move an item in desktop mode (or explicitly switch to grid mode) => In mobile, it's perfect: we have things unimpacted by the desktop grid mode (we have the same options as when it is not in grid mode) and we can switch item order. Now: if the desktop uses grid mode, you have less options in mobile than what you have in normal mode... the users strangely cannot resize their mobile columns/items as in the new normal flow. So he switches back to col mode and make a perfect mobile view... destroying the desktop view.

cc @rdeodoo

robinlej commented:
@qsm-odoo updates done (mainly reordering the commits).

Sorry for the rebase + push-force. I'll be more careful about that in the future!

> (Note: small bug even in website, this goes to "Custom"/"None" only after a new column is clicked, which is strange... should not it update immediately (maybe it is a generic bug about column resizing not update the UI? which would actually reveal more problems I think if that's that) ?)

It's a bug you can already spot on the runbot in master (maybe other versions too, I didn't check) if you:
- Drop a columns snippet
- Resize to 2 or remove one column by clicking on the trash icon
- Copy one column by clicking on the "Duplicate container" icon
=> it still shows "2" until you click on another column.

> Second bug (tested on Firefox, confirmed it does occur only in this branch):

I can't reproduce it with Ctrl + Shift + Print screen, but I do have a traceback at that point with Alt + Tab. :thinking: 

I guess I'll wait more info about the future of this task before trying to fix this.

Finally, concerning the fact that I should have voiced concerns sooner, I'm going to put that on the fact that it was one of my first tasks and I worked on it in the first 2 months, so it didn't strike me. Here again, I'll be more careful in the future. :wink:

qsm-odoo commented:
> I'm going to put that on the fact that it was one of my first tasks and I worked on it in the first 2 months

@robinlej Yes, of course, that's why I am telling it to you :) 

Thanks for the update, but yes: best to wait for additional discussions with PDE now :+1:

robinlej commented:
@qsm-odoo I'm sending this back to testing, but FYI when you'll be reviewing this again:

- I added a test on its own commit (+118). Which partly explains why it might look like there are as many lines as before.
- The main commit is actually +358 / -107. I removed a big chunk of it (± 100 lines) that took care of swapping the offsets when changing the order of 2 items, as it's not actually part of the specs and, well, that was big. I can easily add it back if deemed useful enough after all. (I first added it thinking that offsets are layout-dependent, not item-dependent, so if you add an offset on the 1st item then swap it with the 2nd, you probably want to keep the offset on the element that is visually the 1st. Particularly, it handles the reorder of the 1st and 2nd items on a 5-column layout and it handles the offsets on mobile if you reorder on desktop and don't check afterwards what the layout looks like on mobile. Not sure it's worth the diff, and it's not the current behavior anyway.)

robinlej commented:
Rebased; I hadn't seen that the runbot had failed on my last rebase.

robinlej commented:
@qsm-odoo Where do you stand on this PR? :) (Just to know if it's still on the radar.)

qsm-odoo commented:
@robinlej It's in my queue to be merged before the freeze :wink:

qsm-odoo commented on addons/web_editor/static/src/js/editor/snippets.options.js:
```diff
@@ -3906,19 +3906,6 @@ const SnippetOptionWidget = Widget.extend({
-            // The arrows are not displayed if the target is in a grid and if
-            // not in mobile view.
-            const isMobileView = weUtils.isMobileView(this.$target[0]);
-            if (this.$target[0].classList.contains('o_grid_item') && !isMobileView) {
```
This move does not work, grid items arrow are now always visible :grimacing: And click on them crashes...

robinlej replied:
Done!

qsm-odoo commented on addons/website/static/tests/tours/column_layout.js:
```diff
@@ -0,0 +1,107 @@
+const columnCountOptSelector = ".snippet-option-layout_column we-select[data-name='column_count_opt']";
+const columnsSnippetRow = "iframe .s_three_columns .row";
+
+wTourUtils.registerWebsitePreviewTour("website_update_column_count", {
```
Squash this in the previous commit :+1:

qsm-odoo reviewed (changes requested):
Was going for the merge but quickly found a bug, could you check and I'll look again afterwards? Thanks :+1:

robinlej commented:
@qsm-odoo This should be fixed.
I also changed all uses of `(?<!\S)` as regex lookbehinds are not compatible with Safari < 16.4 ([see this issue](https://github.com/odoo/odoo/pull/134658)).
But I left `(?!\S)`: that's a lookahead and it has been supported forever ([caniuse](https://caniuse.com/mdn-javascript_regular_expressions_lookahead_assertion)).

qsm-odoo commented on addons/web_editor/static/src/js/editor/snippets.options.js:
```diff
@@ -4614,13 +4601,11 @@ registry.sizing = SnippetOptionWidget.extend({
-                const showRight = await this._computeWidgetVisibility('move_right_opt');
-                moveLeftArrowEl.classList.toggle('d-none', !showLeft);
-                moveRightArrowEl.classList.toggle('d-none', !showRight);
+                moveLeftArrowEl.classList.add("d-none");
+                moveRightArrowEl.classList.add("d-none");
```
Sorry but I don't understand this change. Maybe it was already broken before but before this said: "when in mode grid, the visibility of the left/right arrow depends on `_computeWidgetVisibility` (which checks if mobile)". Now this says "When in grid mode and not in mobile view, the left/right arrow should be hidden"... but it does not seem to say that it should be visible in mobile?

At this point of your PR, this is fully bugged: switching to mobile/desktop in grid mode/col mode is very random.

Sometimes I am getting this:
![image](https://github.com/odoo/odoo/assets/10338094/fb064437-fda9-4c43-b2d2-64947c080c42)
Sometimes this:
![image](https://github.com/odoo/odoo/assets/10338094/e75d26cd-fddd-4d14-b859-b1e14362a3dc)

Just by switching to mobile mode/desktop. Maybe it was the case before your PR I don't know but this code seems weird. If broken before your PR, maybe this requires a stable fix before merging moving code around here :grimacing:

robinlej replied:
> before this said: "when in mode grid, the visibility of the left/right arrow depends on `_computeWidgetVisibility` (which checks if mobile)". Now this says "When in grid mode and not in mobile view, the left/right arrow should be hidden"... but it does not seem to say that it should be visible in mobile?

Hmm, so I was going to say "it's visible by default, this only checks if it should override the visibility", but actually no. This whole condition seems unnecessary if I keep the check in the `_computeWidgetVisibility`.
The visibility depends on the `snippetMove`'s `_computeWidgetVisibility`, which is triggered earlier (through a call to `snippet_option_update` in `_onMobilePreviewClick`).
By the time we arrive here, everything has been done already. This really seems superfluous. I'll remove it.

As for the bug, it also happens from 16.0 on. I'll see how to solve it.

robinlej commented:
@qsm-odoo
- The bug should be corrected there: https://github.com/odoo/odoo/pull/138556

As for here, in my [before last push](https://github.com/odoo/odoo/compare/5dfcca4b8f3267c0f8fc791a68429243e4335401..3b41cebbf46841e3b1a1193c2e6b2196f885be6f):
- I removed the `isGrid` check for move arrows and added back the check for grid mode in SnippetMove's `_computeWidgetVisibility`
- I corrected another bug which happened after (1) modifying the order on mobile and (2) modifying the column count with a delta > 1. Since all columns are now added at once to avoid a flickering effect, the mobile order wasn't adapted. Following this change, I also added a step to check that in the test, and I took the opportunity to slightly rearrange the steps order.

sobo-odoo commented on addons/web_editor/static/src/js/editor/snippets.options.js:
```diff
@@ -4597,38 +4578,24 @@ registry.sizing = SnippetOptionWidget.extend({
             for (const handleEl of this.$handles) {
                 const isGridHandle = handleEl.classList.contains('o_grid_handle');
                 handleEl.classList.toggle('d-none', isGrid ^ isGridHandle);
-                // Disabling the resize if we are in mobile view.
-                const isHorizontalSizing = handleEl.matches('.e, .w');
-                handleEl.classList.toggle('readonly', isMobileView && (isHorizontalSizing || isGridHandle));
```
Sorry to interfere but I was curious (since you talked to me about it) so I took a look :slightly_smiling_face: 

If you remove these lines (at least concerning the grid handles) and these lines above 
```js
            // If the handle has the class 'readonly', don't allow to resize.
            // (For the grid handles when we are in mobile view).
            if (ev.currentTarget.classList.contains('readonly')) {
                return;
            }
```
the background grid is "displayed" in grid mode mobile view (see image) :grimacing: 

It is important to not remove the "return if readonly" because when resizing, ALL the sizing options are called one after the other (so sizing_x, sizing_y and sizing_grid) and this is the return that prevents a specific one to be called (so here, the return prevented to call the `onResize` of grid) :wink: Because you checked `isMobile` before displaying it, it is not visible but the space you see on the image is the background grid and it is updated in the `onResize` of sizing_grid 
I do not confirm what I say at 100% because I did not look deeply but this might explain why the resize in mobile view is a bit unstable (because both `onResize` are called successively) 

![image](https://github.com/odoo/odoo/assets/102796037/2b38a4ec-54af-4480-a2ea-b624d2f4d1c8)

sobo-odoo replied:
My bad, wrong guess, these line are not the cause... forget what I just said :sweat_smile: 
But in any case, the background grid should not be visible :wink:

sobo-odoo replied:
Also (last thing and then I will mind my own business :sweat_smile: ) the vertical sizing doesn't work in grid mode + mobile view (I think because this sizing manages the padding and the padding is set differently in grid mode so it has no impact -> to check)
=> maybe you should hide these handles in this case then ? :thinking: Or if not, then find a way to make both padding techniques coexist ?

robinlej replied:
Padding resizing not working on mobile grid is the spec, but weird that the handles appear. Will check it too! Thanks.

qsm-odoo reviewed (changes requested):
Thank you for this work, it looks clean. I did not try to understand every line of it but it looks nice :+1: I tested for a while and I think the behavior we reached is very nice. A bit odd at first when you are used to the previous one but really nice :+1: 

I added 3 comments, 2 of them requiring action if I am right on them. Can you do that post-merge in another PR? :pray: I will push-force a small diff I made and r+ now. I will also create a task for a bug I found which existed before this... but which is now very visible because of your work: if you start resizing a column and mouseout outside of the iframe (which is very easy to do while in mobile mode, which you could not before your work), going back into the frame afterwards starts dragging again :grimacing:

qsm-odoo reviewed (approved):
I made this small diff: https://github.com/odoo/odoo/compare/65f58c37f124c770440028a693f62845bc3c1929..e7c589de2c391ba5d1b896dcb9ef4643dd1d4349

Also forgot to say: I wanted to try mass mailing but it currently crashes on every click in master so difficult to know what's what. In the end, I guess this mobile vs desktop is not to be worried about in mass mailing since it is not possible to edit the mobile view over there anyway? = is there a possibility that your order classes are applied in mass mailing? If yes, that would be a mistake as they would have an effect during preview but not in the sent mail... to confirm :grimacing: 

@robodoo rebase-ff r+

robodoo commented:
Merge method set to rebase and fast-forward.

robodoo commented:
@robinlej @qsm-odoo staging failed: ci/runbot on 2ec0c56b43922892c7399e63f0d0f67398bfbad5 (view more at https://runbot.odoo.com/runbot/build/52680591)

robinlej commented on addons/web_editor/static/src/scss/web_editor.frontend.scss:
```diff
@@ -32,46 +32,61 @@
     transform: scale(-1);
 }
 
+@include media-breakpoint-down(lg) {
+    @for $i from 0 through 12 {
+        .o_we_mobile_order_#{$i} {
+            order: $i;
+        }
+    }
+}
```
I don't remember why I did it that way, but quickly checking I find that bootstrap 5 has 5 order-classes, while bootstrap 4 has 12. (+ one order-first and one order-last for both)
https://getbootstrap.com/docs/4.0/layout/grid/
> "Includes support for 1 through 12 across all five grid tiers."

https://getbootstrap.com/docs/5.0/layout/columns/
> "Includes support for 1 through 5 across all six grid tiers."

6 orders seems ok-ish for our use case but prone to generate errors (just copying one column in a 6-column design would generate one: probably a rare use case, but far from impossible).
And this also means adding `order-lg-0` everywhere there is an `order-x`, which looks to me like a potential source of bugs (although it would probably be contained, since the `order-lg-0` would only be added or removed with all `order-x` classes).

qsm-odoo replied:
We definitely have more that 6:
![image](https://github.com/odoo/odoo/assets/10338094/d2e5ea2f-994c-4ee3-8362-56dafcb1805f)

I guess it's 6 in default BS5 but 12 in our codebase to support BS4, see:
```scss
       "order": map-merge(
            map-get($utilities, "order"),
            (
                values: $order-array,
            ),
        ),
```

Sure one class instead of 2 (order-lg-0) would be better but it's the bootstrap way. I think it's better to avoid extra CSS rules when we can (even if it's only 12 small ones). Actually, your code would be better handling the removal of `order` classes when switching etc, since we should support bootstrap templates as much as possible.

So, in another PR after this, we should make the switch IMO :+1:

qsm-odoo reviewed (approved):
@robodoo r+
